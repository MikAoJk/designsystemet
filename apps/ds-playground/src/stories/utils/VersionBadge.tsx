import { ReactNode } from 'react';

interface VersionBadgeProps {
  packageName: string;
}

export const VersionBadge = ({ packageName }: VersionBadgeProps): ReactNode => {
  let color = 'green';

  let version = process.env[`@skatteetaten/ds-${packageName}`] ?? '..';
  if (process.env['NODE_ENV'] === 'development') {
    color = 'red';
    version = 'developing';
  }
  if (process.env['STORYBOOK_BUILD_VERSION']?.includes('SNAPSHOT')) {
    color = 'blue';
  }

  const normalizedPackageName = packageName.replaceAll('-', '--');
  const normalizedVersion = version.replaceAll('-', '--').replaceAll('_', '__');

  return (
    <img
      src={`https://img.shields.io/badge/%40skatteetaten%2Fds--${normalizedPackageName}-${normalizedVersion}-${color}?logo=npm`}
      alt={`@skatteetaten/ds-${packageName}@${version}`}
    />
  );
};
