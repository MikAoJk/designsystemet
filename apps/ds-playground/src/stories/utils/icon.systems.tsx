import { SystemSVGPathsAndIcons } from '@skatteetaten/ds-icons';

import {
  findSVGPathElements,
  makeLabelsFromSVGPathElements,
} from './icon.utils';

export const SystemSVGPaths = findSVGPathElements(SystemSVGPathsAndIcons);
export const SystemIconLabels = makeLabelsFromSVGPathElements(SystemSVGPaths);
