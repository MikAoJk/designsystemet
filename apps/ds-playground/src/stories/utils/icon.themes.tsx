import { ThemeSVGPathsAndIcons } from '@skatteetaten/ds-icons';

import {
  findSVGPathElements,
  makeLabelsFromSVGPathElements,
} from './icon.utils';

export const ThemeSVGPaths = findSVGPathElements(ThemeSVGPathsAndIcons);
export const ThemeIconLabels = makeLabelsFromSVGPathElements(ThemeSVGPaths);
