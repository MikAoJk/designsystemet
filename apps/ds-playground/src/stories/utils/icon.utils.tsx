import {
  SystemSVGPathsAndIcons,
  ThemeSVGPathsAndIcons,
} from '@skatteetaten/ds-icons';

export type SVGPathAndIconType =
  | typeof SystemSVGPathsAndIcons
  | typeof ThemeSVGPathsAndIcons;
export type SVGPathAndIconKey = keyof SVGPathAndIconType;

export const makeLabelsFromSVGPathElements = function (
  SVGPathElements: Record<SVGPathAndIconKey, SVGPathAndIconType>
): Record<string, string> {
  const labels: Record<string, string> = {};
  Object.keys(SVGPathElements).forEach((name: string) => {
    labels[name] = name.replace('SVGpath', '');
  });

  return labels;
};

export const findSVGPathElements = function (
  SVGPathAndIconElements: Record<SVGPathAndIconKey, SVGPathAndIconType>
): Record<string, SVGPathAndIconType[SVGPathAndIconKey]> {
  const elements: Record<string, SVGPathAndIconType[SVGPathAndIconKey]> = {};
  Object.keys(SVGPathAndIconElements).forEach((elementName: string) => {
    if (elementName.endsWith('SVGpath')) {
      elements[elementName] =
        SVGPathAndIconElements[elementName as SVGPathAndIconKey];
    }
  });

  return elements;
};
