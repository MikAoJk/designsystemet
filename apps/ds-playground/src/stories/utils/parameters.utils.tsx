import { ArgsTable, Stories, PRIMARY_STORY } from '@storybook/addon-docs';
import type { Parameters } from '@storybook/csf';

type DesignParameters = {
  name: string;
  url: string;
};

export const getParameters = function (
  designParameters: DesignParameters[]
): Parameters {
  return {
    docs: {
      page: (): JSX.Element => {
        return (
          <>
            <Stories title={''} />
            <ArgsTable story={PRIMARY_STORY} />
          </>
        );
      },
    },
    design: designParameters.map((designParameter: DesignParameters) => {
      return {
        name: designParameter.name,
        type: 'figma',
        url: designParameter.url,
      };
    }),
    backgrounds: {
      values: [
        { name: 'Svart', value: '#000' },
        { name: 'Hvit', value: '#fff' },
      ],
    },
  };
};
