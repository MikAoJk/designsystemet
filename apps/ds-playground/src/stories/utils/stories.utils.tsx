export const exampleParameters = {
  docs: {
    source: {
      type: 'code',
      language: 'tsx',
      transform: (source: string): string => {
        return source
          .replace('render: (_args): JSX.Element', 'const Examples = ()')
          .replace(
            'render: function Render(_args): JSX.Element ',
            'const Examples = () => '
          )
          .replace('satisfies Story', '');
      },
    },
  },
  /* _args i en story må finnes slik at props i fanen Controls skjules og er prefikset med underscore for å unngå eslint klaging */
  controls: {
    exclude: /.*/,
  },
};
