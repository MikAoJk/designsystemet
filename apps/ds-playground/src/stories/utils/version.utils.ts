export const getVersion = (
  packageName: string
): {
  major: string;
  minor: string;
  patch: string;
  postfix?: string;
  style?: Record<string, string>;
} => {
  const ver = process.env[`@skatteetaten/${packageName}`] ?? '..';
  const [major, minor, patch] = ver.split('.');
  const majorWithPackageName = packageName ? `${packageName}@${major}` : major;

  const style = {
    color: 'var(--semantic-page-foreground)',
    background: 'var(--semantic-success-background)',
    'font-weight': 'var(--font-weight-bold)',
    'font-size': '13px',
  };

  if (process.env['NODE_ENV'] === 'development') {
    return {
      major: majorWithPackageName,
      minor,
      patch,
      postfix: 'develop',
      style: {
        ...style,
        background: 'var(--semantic-danger-background)',
      },
    };
  }
  if (process.env['STORYBOOK_BUILD_VERSION']?.includes('SNAPSHOT')) {
    return {
      major: majorWithPackageName,
      minor,
      patch,
      postfix: process.env['STORYBOOK_BUILD_VERSION'],
      style: {
        ...style,
        background: 'var(--semantic-warning-background)',
      },
    };
  }

  return {
    major: majorWithPackageName,
    minor,
    patch,

    style: {
      ...style,
    },
  };
};
