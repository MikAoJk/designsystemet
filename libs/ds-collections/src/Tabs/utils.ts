export const valueRegex = /^[\w:.-_]*$/;
