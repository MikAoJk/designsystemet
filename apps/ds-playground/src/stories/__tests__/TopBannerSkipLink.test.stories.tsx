import { dsI18n } from '@skatteetaten/ds-core-utils';
import { expect } from '@storybook/jest';
import { Meta, StoryFn, StoryObj } from '@storybook/react';
import { fireEvent, within } from '@storybook/testing-library';

import { wrapper } from './testUtils/storybook.testing.utils';
// eslint-disable-next-line @nx/enforce-module-boundaries
import { TopBannerSkipLink } from '../../../../../libs/ds-layout/src/TopBannerSkipLink/TopBannerSkipLink';
// eslint-disable-next-line @nx/enforce-module-boundaries
import { TopBannerSkipLinkProps } from '../../../../../libs/ds-layout/src/TopBannerSkipLink/TopBannerSkipLink.types';

const meta = {
  component: TopBannerSkipLink,
  title: 'Tester/TopBannerSkipLink (intern)',
  argTypes: {
    // Baseprops
    key: { table: { disable: true } },
    ref: { table: { disable: true } },
    className: { table: { disable: true } },
    id: { table: { disable: true } },
    lang: { table: { disable: true } },
    'data-testid': { table: { disable: true } },
    // Props
    children: { table: { disable: true } },
    // HTML
    href: { table: { disable: true } },
  },
} satisfies Meta<typeof TopBannerSkipLink>;
export default meta;
type Story = StoryObj<typeof meta>;

const skipLinkText = 'ds_pages:topbanner.SkipLinkText';
const defaultArgs: TopBannerSkipLinkProps = {};

export const WithRef = {
  name: 'With Ref (FA1)',
  args: {
    ref: (instance: HTMLAnchorElement | null): void => {
      if (instance) {
        instance.id = 'dummyIdForwardedFromRef';
      }
    },
  },
  argTypes: {
    ref: { table: { disable: false } },
  },
  parameters: {
    imageSnapshot: { disable: true },
  },
  play: async ({ canvasElement }): Promise<void> => {
    const canvas = within(canvasElement);
    const skipLink = canvas.getByRole('link');
    await expect(skipLink).toHaveAttribute('id', 'dummyIdForwardedFromRef');
  },
} satisfies Story;

export const WithAttributes = {
  name: 'With Attributes (FA2-5)',
  args: {
    id: 'htmlId',
    className: 'dummyClassname',
    lang: 'nb',
    'data-testid': '123ID',
  },
  argTypes: {
    id: { table: { disable: false } },
    className: { table: { disable: false } },
    lang: { table: { disable: false } },
    'data-testid': { table: { disable: false } },
  },
  play: async ({ canvasElement }): Promise<void> => {
    const canvas = within(canvasElement);
    const skipLink = canvas.getByRole('link');
    await expect(skipLink).toHaveAttribute('id', 'htmlId');
    await expect(skipLink).toHaveClass('dummyClassname');
    await expect(skipLink).toHaveAttribute('lang', 'nb');
    await expect(skipLink).toHaveAttribute('data-testid', '123ID');
    await skipLink.focus();
  },
} satisfies Story;

export const Defaults = {
  name: 'Defaults (B2)',
  args: {
    ...defaultArgs,
  },
  argTypes: {
    href: { table: { disable: false } },
    children: { table: { disable: false } },
  },
  parameters: {
    imageSnapshot: {
      focus: `${wrapper} a`,
    },
  },
  play: async ({ canvasElement }): Promise<void> => {
    const canvas = within(canvasElement);
    const skipLink = canvas.getByRole('link', {
      name: dsI18n.t(skipLinkText),
    });
    expect(skipLink).toBeInTheDocument();
    expect(skipLink.tagName).toBe('A');
  },
} satisfies Story;

const mainId = 'hovedinnhold';
const MainTemplate: StoryFn<typeof TopBannerSkipLink> = (args) => (
  <>
    <TopBannerSkipLink {...args} />
    <main id={mainId}>{'Hovedinnholdet p√• siden'}</main>
  </>
);

export const WithHrefAndChildren = {
  render: MainTemplate,
  name: 'With Href And Children (A3 delvis)',
  args: {
    ...defaultArgs,
    href: `#${mainId}`,
    children: 'Snarvei til hovedinnhold',
  },
  argTypes: {
    href: { table: { disable: false } },
    children: { table: { disable: false } },
  },
  parameters: {
    imageSnapshot: { disable: true },
  },
  play: async ({ canvasElement }): Promise<void> => {
    const canvas = within(canvasElement);
    const skipLink = canvas.getByRole('link', {
      name: 'Snarvei til hovedinnhold',
    });
    await expect(skipLink).toBeInTheDocument();
    await expect(skipLink).toHaveAttribute('href', `#${mainId}`);
    await fireEvent.click(skipLink);
    const main = canvas.getByRole('main');
    await expect(main).toHaveFocus();
  },
} satisfies Story;

export const ClickSkipLink = {
  render: MainTemplate,
  name: 'Click SkipLink (A3, C1 delvis)',
  args: {
    ...defaultArgs,
  },
  argTypes: {},
  parameters: {
    imageSnapshot: { disable: true },
  },
  play: async ({ canvasElement }): Promise<void> => {
    const canvas = within(canvasElement);
    const main = canvas.getByRole('main');
    const skipLink = canvas.getByRole('link', {
      name: dsI18n.t(skipLinkText),
    });
    await skipLink.focus();
    await fireEvent.click(skipLink);
    await expect(main).toHaveFocus();
  },
} satisfies Story;
