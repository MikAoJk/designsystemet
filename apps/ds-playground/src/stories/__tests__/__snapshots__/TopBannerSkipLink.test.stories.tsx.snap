// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tester/TopBanner/TopBannerSkipLink (intern) ClickSkipLink play-test 1`] = `
<div data-test-block="true">
  <a href="main"
     class="TopBannerSkipLink_skipLink__5N+l7"
  >
    <span class="TopBannerSkipLink_skipLinkContent__T3QpL">
      <svg class="Icon_systemIcon__1p1Xl Icon_systemIcon_medium__VmCFz TopBannerSkipLink_icon__Xwbwo"
           viewbox="0 0 24 24"
           focusable="false"
           aria-hidden="true"
           role="img"
      >
        <path d="M11.1 1.5h1.8v16.8l5.1-5.2 1.3 1.4L12 22l-7.3-7.6L6 13l5.1 5.2V1.5Z">
        </path>
      </svg>
      Hopp til hovedinnhold
    </span>
  </a>
  <main id="hovedinnhold"
        tabindex="-1"
  >
    Hovedinnholdet på siden
  </main>
</div>
`;

exports[`Tester/TopBanner/TopBannerSkipLink (intern) Defaults play-test 1`] = `
<div data-test-block="true">
  <a href="main"
     class="TopBannerSkipLink_skipLink__5N+l7"
  >
    <span class="TopBannerSkipLink_skipLinkContent__T3QpL">
      <svg class="Icon_systemIcon__1p1Xl Icon_systemIcon_medium__VmCFz TopBannerSkipLink_icon__Xwbwo"
           viewbox="0 0 24 24"
           focusable="false"
           aria-hidden="true"
           role="img"
      >
        <path d="M11.1 1.5h1.8v16.8l5.1-5.2 1.3 1.4L12 22l-7.3-7.6L6 13l5.1 5.2V1.5Z">
        </path>
      </svg>
      Hopp til hovedinnhold
    </span>
  </a>
</div>
`;

exports[`Tester/TopBanner/TopBannerSkipLink (intern) WithAttributes play-test 1`] = `
<div data-test-block="true">
  <a id="htmlId"
     href="main"
     class="TopBannerSkipLink_skipLink__5N+l7 dummyClassname"
     lang="nb"
     data-testid="123ID"
  >
    <span class="TopBannerSkipLink_skipLinkContent__T3QpL">
      <svg class="Icon_systemIcon__1p1Xl Icon_systemIcon_medium__VmCFz TopBannerSkipLink_icon__Xwbwo"
           viewbox="0 0 24 24"
           focusable="false"
           aria-hidden="true"
           role="img"
      >
        <path d="M11.1 1.5h1.8v16.8l5.1-5.2 1.3 1.4L12 22l-7.3-7.6L6 13l5.1 5.2V1.5Z">
        </path>
      </svg>
      Hopp til hovedinnhold
    </span>
  </a>
</div>
`;

exports[`Tester/TopBanner/TopBannerSkipLink (intern) WithRef play-test 1`] = `
<div data-test-block="true">
  <a href="main"
     class="TopBannerSkipLink_skipLink__5N+l7"
     id="dummyIdForwardedFromRef"
  >
    <span class="TopBannerSkipLink_skipLinkContent__T3QpL">
      <svg class="Icon_systemIcon__1p1Xl Icon_systemIcon_medium__VmCFz TopBannerSkipLink_icon__Xwbwo"
           viewbox="0 0 24 24"
           focusable="false"
           aria-hidden="true"
           role="img"
      >
        <path d="M11.1 1.5h1.8v16.8l5.1-5.2 1.3 1.4L12 22l-7.3-7.6L6 13l5.1 5.2V1.5Z">
        </path>
      </svg>
      Hopp til hovedinnhold
    </span>
  </a>
</div>
`;

exports[`Tester/TopBanner/TopBannerSkipLink (intern) WithShadowDom play-test 1`] = `
<div data-test-block="true">
  <skiplink-customelement>
  </skiplink-customelement>
</div>
`;

exports[`Tester/TopBanner/TopBannerSkipLink (intern) WithTargetAndChildren play-test 1`] = `
<div data-test-block="true">
  <a href="#hovedinnhold"
     class="TopBannerSkipLink_skipLink__5N+l7"
  >
    <span class="TopBannerSkipLink_skipLinkContent__T3QpL">
      <svg class="Icon_systemIcon__1p1Xl Icon_systemIcon_medium__VmCFz TopBannerSkipLink_icon__Xwbwo"
           viewbox="0 0 24 24"
           focusable="false"
           aria-hidden="true"
           role="img"
      >
        <path d="M11.1 1.5h1.8v16.8l5.1-5.2 1.3 1.4L12 22l-7.3-7.6L6 13l5.1 5.2V1.5Z">
        </path>
      </svg>
      Snarvei til hovedinnhold
    </span>
  </a>
  <main id="hovedinnhold"
        tabindex="-1"
  >
    Hovedinnholdet på siden
  </main>
</div>
`;
